File: src/DIY/Singleton/Singleton.java
package DIY.Singleton;

public class Singleton {
    public static void main(String[] args) {
        ConfigManager config = ConfigManager.getInstance();
        System.out.println("Current settings: ");
        config.printConfig();

        //change settings
        config.setWidth(1280);
        config.setHeight(720);
        config.setScreenMode("WINDOWED");
        config.setBgmVolume(40);
        config.setSfxVolume(23);

        System.out.println("\nModified settings:");
        config.printConfig();

        config.resetToDefault();
        System.out.println("\nDefault settings:");
        config.printConfig();

    }

}

class ConfigManager{

    private static ConfigManager instance;

    private int width;
    private int height;
    private String screenMode;
    private int bgmVolume;
    private int sfxVolume;
    
    private ConfigManager(){
        resetToDefault();
    }
    
    public static ConfigManager getInstance() {
        if (instance == null){
            instance = new ConfigManager();
        }
        return instance;
    }

    public void resetToDefault(){

        width = DefaultConfig.DEFAULT_WIDTH;
        height = DefaultConfig.DEFAULT_HEIGHT;
        screenMode = DefaultConfig.DEFAULT_SCREENMODE;
        bgmVolume = DefaultConfig.DEFAULT_BGM;
        sfxVolume = DefaultConfig.DEFAULT_SFX;

    }
    public void printConfig(){
        System.out.println("Resolution: " + width + " x " + height);
        System.out.println("Screen mode: " + screenMode);
        System.out.println("BGM Volume: " + bgmVolume);
        System.out.println("SFX Volume: " + sfxVolume);
        
    }
    
    public int getWidth() {
        return width;
    }

    public void setWidth(int width) {
        this.width = width;
    }

    public int getHeight() {
        return height;
    }
    
    public void setHeight(int height) {
        this.height = height;
    }

    public String getScreenMode() {
        return screenMode;
    }

    public void setScreenMode(String screenMode) {
        this.screenMode = screenMode;
    }

    public int getBgmVolume() {
        return bgmVolume;
    }

    public void setBgmVolume(int bgmVolume) {
        this.bgmVolume = bgmVolume;
    }

    public int getSfxVolume() {
        return sfxVolume;
    }

    public void setSfxVolume(int sfxVolume) {
        this.sfxVolume = sfxVolume;
    }

}

class DefaultConfig{

    public static final int DEFAULT_WIDTH = 1920;
    public static final int DEFAULT_HEIGHT = 1080;
    public static final String DEFAULT_SCREENMODE = "FULLSCREEN";
    public static final int DEFAULT_BGM = 100;
    public static final int DEFAULT_SFX = 100;

}

File: src/DIY/Singleton/Singleton.ts
// Singleton.ts

class DefaultConfig {
    static readonly DEFAULT_WIDTH = 1920;
    static readonly DEFAULT_HEIGHT = 1080;
    static readonly DEFAULT_SCREENMODE = "FULLSCREEN";
    static readonly DEFAULT_BGM = 100;
    static readonly DEFAULT_SFX = 100;
}

class ConfigManager {
    private static instance: ConfigManager;

    private width: number;
    private height: number;
    private screenMode: string;
    private bgmVolume: number;
    private sfxVolume: number;

    private constructor() {
        this.resetToDefault();
    }

    public static getInstance(): ConfigManager {
        if (!ConfigManager.instance) {
            ConfigManager.instance = new ConfigManager();
        }
        return ConfigManager.instance;
    }

    public resetToDefault(): void {
        this.width = DefaultConfig.DEFAULT_WIDTH;
        this.height = DefaultConfig.DEFAULT_HEIGHT;
        this.screenMode = DefaultConfig.DEFAULT_SCREENMODE;
        this.bgmVolume = DefaultConfig.DEFAULT_BGM;
        this.sfxVolume = DefaultConfig.DEFAULT_SFX;
    }

    public printConfig(): void {
        console.log(`Resolution: ${this.width} x ${this.height}`);
        console.log(`Screen mode: ${this.screenMode}`);
        console.log(`BGM Volume: ${this.bgmVolume}`);
        console.log(`SFX Volume: ${this.sfxVolume}`);
    }

    getWidth(): number {
        return this.width;
    }

    setWidth(width: number): void {
        this.width = width;
    }

    getHeight(): number {
        return this.height;
    }

    setHeight(height: number): void {
        this.height = height;
    }

    getScreenMode(): string {
        return this.screenMode;
    }

    setScreenMode(screenMode: string): void {
        this.screenMode = screenMode;
    }

    getBgmVolume(): number {
        return this.bgmVolume;
    }

    setBgmVolume(bgmVolume: number): void {
        this.bgmVolume = bgmVolume;
    }

    getSfxVolume(): number {
        return this.sfxVolume;
    }

    setSfxVolume(sfxVolume: number): void {
        this.sfxVolume = sfxVolume;
    }
}

// main
(function main() {
    const config = ConfigManager.getInstance();
    console.log("Current settings: ");
    config.printConfig();

    // Change settings
    config.setWidth(1280);
    config.setHeight(720);
    config.setScreenMode("WINDOWED");
    config.setBgmVolume(40);
    config.setSfxVolume(23);

    console.log("\nModified settings:");
    config.printConfig();

    config.resetToDefault();
    console.log("\nDefault settings:");
    config.printConfig();
})();
